src/Castle.Facilities.AspNet.Mvc/AspNetMvcControllerFactory.cs

```diff
@@ -0,0 +1,118 @@
+
+	using Castle.Facilities.AspNet.Mvc.Exceptions;
+	using Castle.MicroKernel;
+	using Castle.MicroKernel.Lifestyle;
+
+	public class AspNetMvcControllerFactory : IControllerFactory
+	{
+		private const string ContextScopeKey = "castle-windsor-facility-aspnet-mvc-scope";
+
+		private readonly ISupportFacility facilitySupport;
+		private List<Assembly> controllerAssemblies;
+
+		public event Action<IKernel, Type> BeforeControllerResolved;
+		public event Action<IKernel, object> AfterControllerReleased;
+
+		public AspNetMvcControllerFactory(ISupportFacility facilitySupport)
+		{
+			this.facilitySupport = facilitySupport;
+		}
+
+		public IController CreateController(RequestContext requestContext, string controllerName)
+		{
+			var adjustedControllerName = $"{controllerName}Controller";
+			var controllerType = FindControllerType(adjustedControllerName);
+			BeginScopedLifestyleIfRequired();
+			OnBeforeControllerResolved(controllerType);
+			try
+			{
+				return (IController) facilitySupport.Kernel.Resolve(controllerType);
+			}
+			catch (InvalidOperationException err)
+			{
+				throw new LifestyleScopesPotentiallyNotEnabledException("If you are having trouble with scoped lifestyles, please try 'container.AddFacility<AspNetMvcFacility>(x => x.AddControllerAssembly<MvcApplication>().WithLifestyleScopedPerWebRequest())'", err);
+			}
+		}
+
+		public SessionStateBehavior GetControllerSessionBehavior(RequestContext requestContext, string controllerName)
+		{
+			return SessionStateBehavior.Default;
+		}
+
+		public void ReleaseController(IController controller)
+		{
+			facilitySupport.Kernel.ReleaseComponent(controller);
+			DisposeScopedLifestyleIfRequired();
+			OnAfterControllerReleased(controller);
+		}
+
+		public void AddControllerAssemblies(List<Assembly> assemblies)
+		{
+			this.controllerAssemblies = assemblies;
+		}
+
+		private Type FindControllerType(string controllerName)
+		{
+			foreach (var type in controllerAssemblies.SelectMany(x => x.ExportedTypes))
+			{
+				if (type.Name == controllerName)
+				{
+					return type;
+				}
+			}
+
+			throw new MissingControllerRegistrationException($"We could not locate the controller '{controllerName}', did you forget to register it?");
+		}
+
+		protected virtual void OnAfterControllerReleased(object controllerInstance)
+		{
+			AfterControllerReleased?.Invoke(facilitySupport.Kernel, controllerInstance);
+		}
+
+		protected virtual void OnBeforeControllerResolved(Type controllerType)
+		{
+			BeforeControllerResolved?.Invoke(facilitySupport.Kernel, controllerType);
+		}
+
+		private void BeginScopedLifestyleIfRequired()
+		{
+			if (facilitySupport.AutoCreateLifestyleScopes)
+			{
+				HttpContext.Current.Items[ContextScopeKey] = facilitySupport.Kernel.BeginScope();
+			}
+		}
+
+		private void DisposeScopedLifestyleIfRequired()
+		{
+			if (facilitySupport.AutoCreateLifestyleScopes)
+			{
+				(HttpContext.Current.Items[ContextScopeKey] as IDisposable)?.Dispose();
```

Comment:
It might be a good idea to remove the item from the HttpContext so the current state of things is clear if debugging.

Reply:
Good spot, done for both mvc and webapi.